Update AS::Notifications::Instrumenter#instrument

alimi opened the pull request:
### Summary

`ActiveSupport::Notifications` provides a powerful API for instrumentation by using a publish-subscribe pattern. We could leverage this for more use cases by opening up `ActiveSupport::Notifications` messaging for things other than instrumentation. To make that happen, I opened this PR to update `ActiveSupport::Notifications::Instrumenter#instrument` so passing a block is optional. This will let us use `#instrument` for general messaging in addition to instrumentation.

/cc @tenderlove @eileencodes

alimi committed:
Update AS::Notifications::Instrumenter#instrument

  * Update #instrument to make passing a block optional. This will let users
    leverage #instrument for messaging in addition to instrumentation.

eileencodes commented on activesupport/lib/active_support/notifications/instrumenter.rb:
```diff
@@ -20,7 +20,7 @@ def instrument(name, payload = {})
         listeners_state = start name, payload
         begin
-          yield payload
+          yield payload if block_given?
```
Can you update the docs above this method?

eileencodes reviewed (changes requested):
Can you update the docs and add a changelog?

alimi committed:
Update CHANGELOG and docs
with change to ActiveSupport::Notifications::Instrumenter#instrument

alimi committed:
Update instrumentation guides w/optional block

alimi commented:
I updated the changelog and docs in https://github.com/rails/rails/pull/35705/commits/aaf89cdea7105ae7789483cec2d786cfa1cef7f9 and updated the guides in https://github.com/rails/rails/pull/35705/commits/7237c7be74b09cd951ff000afb4087313e5c0025.
